### ISSUE TITLE
Tauri Documentation Ecosystem Evolution

### ISSUE THREAD

**Issue by lorenzolewis** [tauri-apps/tauri-docs#1040]
# TL;DR

Over the past year Tauri has expanded in terms of user base as well as functionality. With the upcoming mobile launch we will have even more need for clean documentation. This is the beginning of rolling our own custom documentation and guide platform backed by Astro. We'd like to have these efforts live by mid-2023.

# Overview

- The main issue is that we're limited by the performance of Docusaurus's stack (which is what's used to back the site now) due to the amount of time it takes to build a site (https://github.com/facebook/docusaurus/issues/4765).
- We've also made decisions that have limited us (using markdown to back the JS API docs) or added complexity and fragility to our CI (such as MDX breaking i18n commonly).
- Lastly we want enable technical guides (such as the articles on https://tauri.app/v1/guides/features/ or the upcoming Tauri Plugin documentation) to be closer to the source code to make it easier to maintain and update for those working in the code while also allowing mirroring them on the website

We'll start by putting together a team on Discord (message LorenzoLewis#2572 or JonasKruckenberg#2243 on discord if interested in joining) to action upon this vision. We'll roughly follow the roadmap below but that is expected to morph and change over time as we dig into the details.

# Skills Needed

- [ ] Rust & JS AST expertise/willingness to dive in
- [ ] Design/UI/UX
- [ ] Astro Component Development
- [ ] SEO & Accessibility for Astro
- [ ] GitHub and Netlify CI/CD to integrate repos together (possibly Search as well)
- [ ] Docs writers (a bit later down the line after we setup the infra)

# Roadmap

**Phase 1**

- [ ] 1. Port blog, about pages, guides & references, JS API, Config and CLI to Astro
  - [x] 1. Build the URL structure in Astro as outlined here: [https://hackmd.io/@lorenzo-lewis/HyQLZrMFs](https://hackmd.io/@lorenzo-lewis/HyQLZrMFs)
    - Complete via 64096128b783b3e8ffa290f6dc7571eda75c70d6 
  - [ ] 2. Create the Astro layouts for markdown files
  - [ ] 3. Build out the i18n integration with Crowdin (notes: https://hackmd.io/@lorenzo-lewis/ryf5ovQFo)
- [ ] 2. Design & Test new site UI/UX
  - [ ] 1. Design and build base layout components (navbar, footer, reusable sidebars, etc.)
  - [ ] 2. Design and build Astro layouts and components for the various page types:
    - [ ] 1. Pages: Non-Markdown pages (such as index)
    - [ ] 2. Docs: Technical guides driven by markdown (the “Guides” and “References” section on the current site)
    - [ ] 3. Blog posts
    - [ ] 4. JS APIs driven by Markdown
    - [ ] 5. Tauri Config driven by Markdown
    - [ ] 6. Tauri CLI driven by Markdown

**Phase 2**

- [ ] 1. Move guides to upstream repos (such as `tauri-apps/tauri` and `tauri-apps/plugin-workspace`)
  - [ ] 1. Define and build CI/CD integration with upstream maintainers
  - [ ] 2. Create mdBook projects in upstream repos and move respective content upstream
  - [ ] 3. Create a standard format/template for technical guides and then write content to reflect that. Note that the scope of these guides should be to cover all of the top-level modules and their functionality
- [ ] 2. Rebuild API pages as Astro pages
  - [ ] 1. Tauri config: Design and build page driven off of JSON AST
  - [ ] 2. Tauri CLI: Might be fine leaving it as-is (driven off of markdown), need to do research here to see what improvements we could gain
  - [ ] 3. JavaScript/TypeScript API: Design and build to be driven off of AST generated from TypeDoc into Astro components
  - [ ] 4. Rust API: Design and build to be driven off of AST generated by `rustdoc-json`. Doesn’t need to be as detailed as [[docs.rs](http://docs.rs/)](http://docs.rs) in its design but should be focused on approachability from those not as familiar with the Rust ecosystem. (Notes: https://hackmd.io/@lorenzo-lewis/Bk_pqUEto)

# High-Level Inter-Repo Architecture

```mermaid
graph LR
  F[Render Docs]
		
subgraph tauri-apps/tauri-docs
  D[Markdown Guides] --> F
  P --> F
  O --> F
  F --> H[tauri.app hosts...]
  H --> N[Guides]
  H --> I[Core JS API]
  H --> J[Core Rust API]
  H --> M[Tauri Config Schema]
  H --> K[Plugin API]
  H --> Z[Plugin Guides]
end

subgraph tauri-apps/tauri
  Q[Core Feature Markdown Guides] --> F
  L[Tauri Config Schema JSON] --> F
  B[Core JS/TS code] --> P[Generated AST JSON via TypeDoc]
  A[Core Rust code] --> O[Generated AST JSON via rustdoc-json]
end

subgraph tauri-apps/plugins-workspace
	  C[Plugin Rust code] --> O
		R[Plugin Markdown Guides] --> F
end
```

# Big List of Ideas

This is just a way for us to organise our thoughts, the roadmap above should be the steps we follow.

## High priority

- [ ] Way to incorporate technical details (inline API references) into guides
- [ ] Improve ability and flexibility for i18n
- [ ] Only use pure MD for guides, no MDX (i18n chokes on it)
- [ ] Keep sources close to the code but allow easily building and previewing the output (so plugin guides in the plugin repo, but you can see a loosely formatted version within that repo powered by MdBook)
- [ ] Search: Need a better and more "plug-and-play" search service. A lot of feedback on how our current search just isn't good from the Tauri Community Survey.

## Medium Priority

- [ ] Inline code highlighting (Rust and JS)
- [ ] A nice way to show different implementations or outputs between platforms where needed (macOS, Windows, Linux, iOS, Android)

## Low Priority / Aspirations

- [ ] Ability to build screenshots for guides in CI pipeline for all deploy targets

# Random Notes

Mainly for me to document random ideas I have, might be useful to someone, might just cause confusion 🤷‍♂️

- Learning framework https://diataxis.fr
- Guides shouldn't be versioned, but API docs should be
- For CI/CD we should do something like `dev.tauri.app` that continually builds guides and APIs throughout the Tauri repos to give a faster feedback loop and see if things break
- For MD guides, how do we reference symbols within that repo so we don't have to link out to the site directly? 
- Tab group representation in markdown: https://hackmd.io/@lorenzo-lewis/B1_FXrMFi
- A11Y: https://github.com/pa11y/pa11y
- Our design system is based on this [Figma file](https://www.figma.com/file/CKTllB7R52D080yDHjno92/Tauri-Brand-Guidelines?node-id=953%3A1977&t=zdCbYb4lqSzUqv2d-0)
- Would like to adapt to provide more continuity between dark and light modes as well as more flexible colour usage (accents, warning, info, etc.)
- Revise search system, this can most likely be done on the current instance prior to rolling out the Astro site
- Alternative i18n: https://github.com/Alexandre-Fernandez/astro-i18n

CC @JonasKruckenberg
